datascience_eda/datascience_eda.py

```diff
@@ -74,31 +168,293 @@ def explore_clustering(
+
+        if include_silhouette:
+            s_visualizer = SilhouetteVisualizer(kmeans, colors="yellowbrick")
+            s_visualizer.fit(x)  # Fit the data to the visualizer
+            s_visualizer.show()
+            plt.clf()
+            plt.close()
+            silhouette_plots.append(s_visualizer)
+        else:
+            silhouette_plots.append(None)
+
+        # PCA plots
+        if include_silhouette:
+            labels = kmeans.fit_predict(x)
+            pca_fig = plot_pca_clusters(x, labels)
+            pca_plots.append(pca_fig)
+        else:
+            pca_plots.append(None)
+
+    results["Silhouette"] = silhouette_plots
+    results["PCA"] = pca_plots
 
-    # plot PCA clusters
+    return results
 
-    # add all plots to result dictionary
 
-    return result
+def explore_DBSCAN_clustering(
+    df,
+    num_cols=None,
+    metric="euclidean",
+    eps=[0.5],
+    min_samples=[5],
+    include_silhouette=True,
+    include_PCA=True,
+):
+    """fit and plot DBSCAN clustering algorithms
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        the dataset, should be transformed with StandardScaler
+    num_cols : list, optional
+        list of numeric column names, in case of None, get all numeric columns
+    metric : str, optional
+        metric, by default "euclidean"
+    eps : list, optional
+        list of eps hyperparams, by default [0.5]
+    min_samples: list, optional
+        list of min_samples hyperparams, by default [5]
+    include_silhouette : bool, optional
+        whether Silhouette plots should be generated, by default True
+    include_PCA : bool, optional
+        whether PCA plots should be generated, by default True
+
+    Returns
+    -------
+    Tuple
+        list
+            a list of n_clusters values returned by DBSCAN models
+        dict
+            a dictionary with key=type of plot, value=list of plots
+
+    Examples
+    -------
+    >>> original_df = pd.read_csv(currentdir + "/data/menu.csv")
+    >>> numeric_features = eda.get_numeric_columns(original_df)
+    >>> drop_features = []
+    >>> numeric_transformer = make_pipeline(SimpleImputer(), StandardScaler())
+    >>> preprocessor = make_column_transformer(
+    >>>     (numeric_transformer, numeric_features), ("drop", drop_features)
+    >>> )
+    >>> df = pd.DataFrame(
+    >>>     data=preprocessor.fit_transform(original_df), columns=numeric_features
+    >>> )
+    >>> n_clusters, dbscan_plots = explore_DBSCAN_clusterting(df)
+    """
+    if num_cols is None:
+        num_cols = get_numeric_columns(df)
+    else:
+        _verify_numeric_cols(df, num_cols)
+
+    x = df[num_cols]
+
+    results = {}
+    n_clusters = []
+
+    s_plots = []
+    pca_plots = []
+
+    print("------------------------")
+    print("DBSCAN CLUSTERING")
+    print("------------------------")
+
+    for e in eps:
+        for ms in min_samples:
+            dbscan = DBSCAN(eps=e, min_samples=ms, metric=metric)
```

Comment:
Could you please let me know if the there would be an error message thrown when metric is not of string type?

Reply:
the metric is verified earlier in the function